proposals/add-deque-struct.md

```diff
@@ -0,0 +1,153 @@
+# Proposal: Add a `Deque` struct to the stdlib
+
+## Goals
+
+The goals of this document are
+
+1) Explain what problems we want to solve with `Deque`, and what problems we
+   don't want to solve.
+2) Explain how the internal storage of this double-ended
+   queue is managed. Give explored alternatives.
+3) Propose an API which can be used as a "todo list"
+   for contributors to implement.
+
+## 1 - Why we want to add a `Deque` struct to the stdlib
+
+All major languages have a double-ended queue in their stdlib.
+And they are useful for a number of reasons.
+The major one is that it allows users to work with
+FIFO (first-in first-out) and LIFO (last-in first-out) queues,
+which are very common in programming.
+
+For context, we are still at the start of Mojo's stdlib.
+Given the number of maintainers,
+we want to create a new data structure which
+doesn't add too much code and complexity to the codebase.
+
+With those constraints we give the following goals:
+
+1) Provide a data structure which allows users to
+   implement a fast FIFO and LIFO
+2) Has a familiar api with already existing Mojo
+   collections (`List`, `InlineList`, `String`, `VariadicList`, ...)
+3) Has a familiar api to Python users
+4) Can be efficiently converted from and into a `List`.
+5) Has fast access to its elements (cache-friendly).
```

Comment:
IMO 4) should not be a goal for the data structure we will call `deque` (or `Deque`). Neither CPP or Python makes this promise, and those are the languages/stdlibs closest to Mojo, and compatibility with Python is a stated goal (although, I admit, the level of this compatibility is not well defined). CPP does promise 5), and I think it's reasonable to aim for it (but maybe not right away).